### PR TITLE
Add option to disable caching for HTML content

### PR DESCRIPTION
**Pull request by chenyhd**
## Type of Change

- **Build Scripts:** add `Cache-Control: no-cache` for HTML source to improve the performance

### What should this PR do?
Global performance improves, add no-cache for HTML

### What are the acceptance criteria?

These changes will add `Cache-Control: no-cache` for website HTML

#### Environmental preparation
1. Nginx website nginx location config
```
listen      80;
# ...
 set                     $base /var/www/example.com;
 root                    $base/public;
 location / {
     try_files $uri $uri/ /index.html;
 }
```
2. Create index.html in `/var/www/example.com/public/cache-test` and input anything like:
```
<h1>Caching Test</h1>
```
3. In order to prevent the influence of successive operations, please use incognito mode

#### previous behavior
1. When you visit `localhost/cache-test` with Chrome first time
You will get `200` status code
2. `F5` this page
You will get `304` status code
3. **When you open a new tab and  visit `localhost/cache-test`**
You will get `200` status code with a cache
image

#### Current behavior
Nginx config change to
```
listen      80;
# ...
 set                     $base /var/www/example.com;
 root                    $base/public;

# This commit chages
location ~* \.(?:html)$ {
    add_header Cache-Control "no-cache";
    access_log off;
}


 location / {
     try_files $uri $uri/ /index.html;
 }
```

1. When you visit `localhost/cache-test` with Chrome first time
You will get `200` status code
2. `F5` this page
You will get `304` status code
3. **When you open a new tab and  visit `localhost/cache-test`**
You will get `304` status code

UI Changes:

Before:
image

After:
image
image

### CODE REVIEW

**Comment by chenyhd** [comment]
### Add some information

- [How no-cache work](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#force_revalidation)
- [Chrome behaviors](https://stackoverflow.com/questions/26166433/how-to-prevent-request-that-returns-304#:~:text=Naturally%2C%20all%20resources%20were%20served%20with%20200%20codes%20on%20first%20page%20load.%20Thereafter%2C%20testing%20in%20Chrome%20or%20Firefox%20installs%20with%20default%20settings%2C%20I%20observed%20that%3A)

> 
- If you refresh the page via the F5 key or the Refresh button, the page and all resources revalidate (i.e. a request is made to the server for each resource and a 304 returned).
- If you return to the page via a link or entering the URL into the URL bar in a new tab, then no revalidation is done (i.e. no requests are made).
- In Chrome, if you refresh the page by selecting the URL bar and pressing Enter, the page itself revalidates but no other resources do. In Firefox, neither the page nor resources revalidate.

**Comment by MattIPv4** [comment]
👋 I'm not sure what you're trying to achieve here, but setting a no-cache header is not going to improve cache performance -- it will in fact do the opposite, it will tell the browser to never cache the file locally and always make a fresh request to the origin (NGINX) to get the file.

**Comment by MattIPv4** [comment]
If you want to be able to customise the cache lifetime of HTML files, instead of just adding a checkbox that forces them to never be cached, I'd suggest mirroring how we set cache lifetimes for other asset types already.

**Comment by chenyhd** [comment]
> I'm not sure what you're trying to achieve here

A: When we deploy a new frontend version the client will receive the latest changes in time

> but setting a no-cache header is not going to improve cache performance -- it will in fact do the opposite, 

A: My understanding is skewed, I thought it was related to "performance" rather than "improving performance". 

> it will tell the browser to never cache the file locally and always make a fresh request to the origin (NGINX) to get the file.

> If you want to be able to customise the cache lifetime of HTML files, instead of just adding a checkbox that forces them to never be cached, I'd suggest mirroring how we set cache lifetimes for other asset types already.

Let me provide some example:

1.[DigitalOcean NGINXConfig](https://www.digitalocean.com/community/tools/nginx)
`cache-control: max-age=0`
image

2.[Google](https://www.google.com/)
`cache-control: private, max-age=0`
image


3.[Amazon](https://www.amazon.com/)
`cache-control: no-cache`
image

HTML is special, If HTML is changed that means related CSS and JS need to be refreshed, If HTML isn't changed that means get CSS and JS from the local cache and follow the expiration setting

`max-age=0` and `no-cache` have the same effect

**Comment by MattIPv4** [comment]
Ah okay, I see what you are trying to do now -- could we name this feature "Disable HTML caching" instead then?

**Comment by chenyhd** [comment]
I have changed it to "Disable HTML caching"
image

File: src/nginxconfig/i18n/de/templates/global_sections/performance.js [hunk 1]
```diff
@@ -27,6 +27,8 @@ THE SOFTWARE.
 import common from '../../common';
 
 export default {
+    disableHtmlCaching: 'Disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/de/templates/global_sections/performance.js:30]:
```suggestion
    disableHtmlCaching: 'Disable HTML caching', // TODO: translate
```

File: src/nginxconfig/i18n/de/templates/global_sections/performance.js [hunk 2]
```diff
@@ -27,6 +27,8 @@ THE SOFTWARE.
...
+    enableChangesTakeEffect: 'disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/de/templates/global_sections/performance.js:31]:
```suggestion
    enableDisableHtmlCaching: 'disable HTML caching', // TODO: translate
```

File: src/nginxconfig/i18n/es/templates/global_sections/performance.js [hunk 3]
```diff
@@ -26,7 +26,9 @@ THE SOFTWARE.
 
 import common from '../../common';
 
-export default {
+export default {    
+    disableHtmlCaching: 'Disable HTML caching',
+    enableChangesTakeEffect: 'disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/es/templates/global_sections/performance.js:31]:
```suggestion
    enableDisableHtmlCaching: 'disable HTML caching', // TODO: translate
```

File: src/nginxconfig/i18n/pt-br/templates/global_sections/performance.js [hunk 4]
```diff
@@ -27,6 +27,8 @@ THE SOFTWARE.
 import common from '../../common';
 
 export default {
+    disableHtmlCaching: 'Disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/pt-br/templates/global_sections/performance.js:30]:
```suggestion
export default {
    disableHtmlCaching: 'Disable HTML caching', // TODO: translate
```

File: src/nginxconfig/i18n/zh-tw/templates/global_sections/performance.js [hunk 5]
```diff
@@ -27,6 +27,8 @@ THE SOFTWARE.
 import common from '../../common';
 
 export default {
+    disableHtmlCaching: 'Disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/zh-tw/templates/global_sections/performance.js:30]:
```suggestion
    disableHtmlCaching: 'Disable HTML caching', // TODO: translate
```

File: src/nginxconfig/i18n/zh-tw/templates/global_sections/performance.js [hunk 6]
```diff
@@ -27,6 +27,8 @@ THE SOFTWARE.
...
+    enableChangesTakeEffect: 'disable HTML caching',
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/zh-tw/templates/global_sections/performance.js:31]:
```suggestion
    enableDisableHtmlCaching: 'disable HTML caching', // TODO: translate
```

File: src/nginxconfig/templates/global_sections/performance.vue [hunk 7]
```diff
@@ -26,6 +26,25 @@ THE SOFTWARE.
 
 <template>
     <div>
+
+
```
Comment (MattIPv4) [inline src/nginxconfig/templates/global_sections/performance.vue:30]:
```suggestion
```

File: src/nginxconfig/templates/global_sections/performance.vue [hunk 8]
```diff
@@ -26,6 +26,25 @@ THE SOFTWARE.
...
+        <div class="field is-horizontal">
+            <div class="field-label">
+                <label class="label">{{ $t('templates.globalSections.performance.disableHtmlCaching') }}</label>
+            </div>
+            <div class="field-body">
+                <div class="field">
+                    <div :class="`control${disableHtmlCachingChanged ? ' is-changed' : ''}`">
+                        <div class="checkbox">
+                            <PrettyCheck v-model="disableHtmlCaching" class="p-default p-curve p-fill p-icon">
+                                {{ $t('templates.globalSections.performance.enableChangesTakeEffect') }}
```
Comment (MattIPv4) [inline src/nginxconfig/templates/global_sections/performance.vue:40]:
```suggestion
                                {{ $t('templates.globalSections.performance.enableDisableHtmlCaching') }}
```

File: src/nginxconfig/templates/global_sections/performance.vue [hunk 9]
```diff
@@ -127,23 +146,8 @@ THE SOFTWARE.
...
-                        <input
-                            v-model="fontsExpiration"
-                            class="input"
-                            type="text"
-                            :placeholder="$props.data.fontsExpiration.default"
-                        />
-                    </div>
-                </div>
-            </div>
-        </div>
```
Comment (MattIPv4) [inline src/nginxconfig/templates/global_sections/performance.vue:146]: Why has this been removed? Please revert

File: src/nginxconfig/templates/global_sections/performance.vue [hunk 10]
```diff
@@ -154,6 +158,10 @@ THE SOFTWARE.
     import PrettyCheck from '../inputs/checkbox';
 
     const defaults = {
+        disableHtmlCaching: {
+            default: true,
```
Comment (MattIPv4) [inline src/nginxconfig/templates/global_sections/performance.vue:162]:
```suggestion
            default: false,
```
To ensure backwards compatibility in the tool, let's set this to false by default

File: src/nginxconfig/util/angular_backwards_compatibility.js [hunk 11]
```diff
@@ -45,6 +45,7 @@ const globalMap = {
 
     python_server:              ['python', 'pythonServer'],
 
+    disableHtmlCaching:         ['performance', 'disableHtmlCaching', oldBool],
```
Comment (MattIPv4) [inline src/nginxconfig/util/angular_backwards_compatibility.js:48]:
```suggestion
```
This doesn't need to be included here as this was never a feature in the Angular version of the tool

File: src/nginxconfig/util/types_extensions.js [hunk 12]
```diff
@@ -37,4 +37,5 @@ export const extensions = {
         'docx?|dotx?|docm|dotm|' +
         'xlsx?|xltx?|xlsm|xltm|' +
         'pptx?|potx?|pptm|potm|ppsx?',
+    html: 'html',
```
Comment (MattIPv4) [inline src/nginxconfig/util/types_extensions.js:40]:
```suggestion
    html: 'html?',
```
Allow `html` or `htm`

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
:wave: Functionality wise this is looking great, just some quick cleanup notes for the PR 💙 

---

Please could you make sure the copyright year is 2022 in all the files you've updated?

**Comment by chenyhd** [comment]
Done.

**Review by MattIPv4 — APPROVED** [review]
LGTM, tyvm

**Comment by MattIPv4** [comment]
Hey, @chenyhd - Thanks a ton for this sweet PR! 😄

Would you please shoot me an email when you get a chance?
mcowley at digitalocean dot com 🎉

**Pull request merged**